src/event/ngx_event_openssl_cache.c

```diff
@@ -0,0 +1,326 @@
+    ngx_str_t *path, ngx_ssl_cache_key_t *id);
+static ngx_ssl_cache_node_t *ngx_ssl_cache_lookup(ngx_ssl_cache_t *cache,
+    ngx_ssl_cache_type_t *type, ngx_ssl_cache_key_t *id, uint32_t hash);
+
+static void *ngx_openssl_cache_create_conf(ngx_cycle_t *cycle);
+static void ngx_ssl_cache_cleanup(void *data);
+static void ngx_ssl_cache_node_insert(ngx_rbtree_node_t *temp,
+    ngx_rbtree_node_t *node, ngx_rbtree_node_t *sentinel);
+
+
+static ngx_core_module_t  ngx_openssl_cache_module_ctx = {
+    ngx_string("openssl_cache"),
+    ngx_openssl_cache_create_conf,
+    NULL
+};
+
+
+ngx_module_t  ngx_openssl_cache_module = {
+    NGX_MODULE_V1,
+    &ngx_openssl_cache_module_ctx,         /* module context */
+    NULL,                                  /* module directives */
+    NGX_CORE_MODULE,                       /* module type */
+    NULL,                                  /* init master */
+    NULL,                                  /* init module */
+    NULL,                                  /* init process */
+    NULL,                                  /* init thread */
+    NULL,                                  /* exit thread */
+    NULL,                                  /* exit process */
+    NULL,                                  /* exit master */
+    NGX_MODULE_V1_PADDING
+};
+
+
+static ngx_ssl_cache_type_t  ngx_ssl_cache_types[] = {
+
+};
+
+
+void *
+ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
+    ngx_str_t *path, void *data)
+{
+    void                  *value;
+    uint32_t               hash;
+    ngx_ssl_cache_t       *cache;
+    ngx_ssl_cache_key_t    id;
+    ngx_ssl_cache_type_t  *type;
+    ngx_ssl_cache_node_t  *cn;
+
+    if (ngx_ssl_cache_init_key(cf->pool, index, path, &id) != NGX_OK) {
+        return NULL;
+    }
+
+    cache = (ngx_ssl_cache_t *) ngx_get_conf(cf->cycle->conf_ctx,
+                                             ngx_openssl_cache_module);
+
+    type = &ngx_ssl_cache_types[index];
+    hash = ngx_murmur_hash2(id.data, id.len);
+    value = NULL;
+
+    cn = ngx_ssl_cache_lookup(cache, type, &id, hash);
+
+    if (cn == NULL) {
+        cn = ngx_palloc(cf->pool, sizeof(ngx_ssl_cache_node_t) + id.len + 1);
+        if (cn == NULL) {
+            return NULL;
+        }
+
+        cn->node.key = hash;
+        cn->id.data = (u_char *)(cn + 1);
+        cn->id.len = id.len;
+        cn->id.type = id.type;
+        cn->type = type;
+        cn->value = NULL;
+
+        ngx_cpystrn(cn->id.data, id.data, id.len + 1);
+
+        ngx_rbtree_insert(&cache->rbtree, &cn->node);
+    }
+
+    /* try to use a reference from the cache */
+
+    if (cn->value != NULL) {
+        value = type->ref(err, cn->value);
+    }
+
+    if (value == NULL) {
+        value = type->create(&id, err, data);
+    }
+
+    if (value != NULL && cn->value == NULL) {
+        /* we have a value and the node needs one; try to reference it */
+        cn->value = type->ref(err, value);
+    }
```

Comment:
What about simplifying this:

```
    if (cn->value == NULL) {
        cn->value = type->create(&id, err, data);
        if (cn->value == NULL) {
            return NULL;
        }
     }
 
    return type->ref(err, cn->value);
```
This code will not create a new element after a ref() error, but I think it should not happen anyway.

Reply:
Thanks for prodding this.  Actually, we can further simplify and streamline this function.
In pseudo-code:
- if (lookup) then return ref;
- allocate/init new node;
- create cn->value or return error;
- insert rbtree
- return ref

Actual patch on top (better viewed with `git diff -w`):

```diff
diff --git a/src/event/ngx_event_openssl_cache.c b/src/event/ngx_event_openssl_cache.c
index f059f88f5..4c9ff1ec7 100644
--- a/src/event/ngx_event_openssl_cache.c
+++ b/src/event/ngx_event_openssl_cache.c
@@ -132,7 +132,6 @@ void *
 ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
     ngx_str_t *path, void *data)
 {
-    void                  *value;
     uint32_t               hash;
     ngx_ssl_cache_t       *cache;
     ngx_ssl_cache_key_t    id;
@@ -148,44 +147,34 @@ ngx_ssl_cache_fetch(ngx_conf_t *cf, ngx_uint_t index, char **err,
 
     type = &ngx_ssl_cache_types[index];
     hash = ngx_murmur_hash2(id.data, id.len);
-    value = NULL;
 
     cn = ngx_ssl_cache_lookup(cache, type, &id, hash);
+    if (cn != NULL) {
+        return type->ref(err, cn->value);
+    }
 
+    cn = ngx_palloc(cf->pool, sizeof(ngx_ssl_cache_node_t) + id.len + 1);
     if (cn == NULL) {
-        cn = ngx_palloc(cf->pool, sizeof(ngx_ssl_cache_node_t) + id.len + 1);
-        if (cn == NULL) {
-            return NULL;
-        }
-
-        cn->node.key = hash;
-        cn->id.data = (u_char *)(cn + 1);
-        cn->id.len = id.len;
-        cn->id.type = id.type;
-        cn->type = type;
-        cn->value = NULL;
-
-        ngx_cpystrn(cn->id.data, id.data, id.len + 1);
-
-        ngx_rbtree_insert(&cache->rbtree, &cn->node);
+        return NULL;
     }
 
-    /* try to use a reference from the cache */
+    cn->node.key = hash;
+    cn->id.data = (u_char *)(cn + 1);
+    cn->id.len = id.len;
+    cn->id.type = id.type;
+    cn->type = type;
+    cn->value = NULL;
 
-    if (cn->value != NULL) {
-        value = type->ref(err, cn->value);
-    }
+    ngx_cpystrn(cn->id.data, id.data, id.len + 1);
 
-    if (value == NULL) {
-        value = type->create(&id, err, data);
+    cn->value = type->create(&id, err, data);
+    if (cn->value == NULL) {
+        return NULL;
     }
 
-    if (value != NULL && cn->value == NULL) {
-        /* we have a value and the node needs one; try to reference it */
-        cn->value = type->ref(err, value);
-    }
+    ngx_rbtree_insert(&cache->rbtree, &cn->node);
 
-    return value;
+    return type->ref(err, cn->value);
 }
 
 
```
The important point is that now we maintain a contract that for a cached node `cn->value != NULL` is always true.
This allows to remove certain checks to simplify code.

Reply:
With the exception of `cn->value = NULL;` looks good.